src/file_watching.jl

```
@@ -1,56 +1,153 @@
-    watched_files::Dict{String,Task}
-    callback::Function
-    FileWatcher(cb::Function) = new(Dict{String,Task}(), cb)
+mutable struct SimpleWatcher
+    callback::Union{Nothing,Function}
+    task::Union{Nothing,Task}
+    sleeptime::Float64
+    filelist::Vector{WatchedFile}
+    SimpleWatcher(fcn::Union{Nothing,Function}=nothing; sleeptime::Float64=0.2) = new(fcn,nothing,max(0.1,sleeptime),Vector{WatchedFile}())
 end
 
 """
-    _file_watcher(filepath, fw)
+    _file_watcher(w::SimpleWatcher)
 
-Helper function that's called asynchronously to act if a file has changed.
-See also [`add_to_filewatcher`](@ref).
+[INTERNAL] Helper function that's spawned as an asynchronous task which
+checks for changed files. Terminates normally upon a `InterruptException`,
+and with a warning for all other exceptions.
 """
-function _file_watcher(filepath::AbstractString, fw::FileWatcher)
-    fileevent = watch_file(filepath, -1) # returns upon a change
-    # if file changed, start new task on this file (s.t. it is still being watched)
-    if !fileevent.renamed && !fileevent.timedout
-        fw.watched_files[filepath] = @async _file_watcher(filepath, fw)
-    else
-        # if renamed or timed out, remove from list of watched files
-        delete!(fw.watched_files, filepath)
+function _file_watcher(w::SimpleWatcher)
+    try
+        while true
+            # checking for changed files
+            changed_files = Vector{String}()
+            foreach(w.filelist) do wf
+                if has_changed(wf)
+                    set_unchanged!(wf)
+                    push!(changed_files, wf.path)
+                end
+            end
+
+            # invoke callback on each changed file (if it's defined), then sleep
+            (w.callback != nothing) && foreach(w.callback, changed_files)
+            sleep(w.sleeptime)
+        end
+    catch EXC
+        if !isa(EXC, InterruptException) # if interruption, normal termination
+            @warn "Exception in hdlr: " EXC
```

probably more than @warn here. you probably need `throw`  as it's an exception after all.

---

The problem is that when the error is thrown, it won't show / have an effect on the main task. So the watcher task fails silently and the live server just does not work anymore. I'll keep the warning for now (with a more concise message, such as "please stop the server"). We'd need a way of terminating the server upon error in this task...?

---

well you can close the server from here? Note that this kind of stuff is why it seems a bit unnecessary to have more than just the HTTP listener be async. But anyway if the global thing is to check and act upon the server then basically everyone should have access to it and verify it's still open and potentially close it upon error.

---

Will look into this in #15 once this is merged.